Fill.py

```diff
@@ -258,6 +258,45 @@ def distribute_items_restrictive(world: MultiWorld) -> None:
     usefulitempool: typing.List[Item] = []
     filleritempool: typing.List[Item] = []
 
+    early_items_count: typing.Dict[typing.Tuple[str, int], int] = {}
+    for player in world.player_ids:
+        for item, count in world.early_items[player].value.items():
+            early_items_count[(item, player)] = count
+    if early_items_count:
+        early_locations: typing.List[Location] = []
+        early_priority_locations: typing.List[Location] = []
+        for loc in reversed(fill_locations):
+            if loc.can_reach(world.state):
+                if loc.progress_type == LocationProgressType.PRIORITY:
+                    early_priority_locations.append(loc)
+                else:
+                    early_locations.append(loc)
+                fill_locations.remove(loc)
+
+        early_prog_items: typing.List[Item] = []
+        early_rest_items: typing.List[Item] = []
+        for item in reversed(itempool):
+            if (item.name, item.player) in early_items_count:
+                if item.advancement:
+                    early_prog_items.append(item)
+                else:
+                    early_rest_items.append(item)
+                itempool.remove(item)
+                early_items_count[(item.name, item.player)] -= 1
+                if early_items_count[(item.name, item.player)] == 0:
+                    del early_items_count[(item.name, item.player)]
+        fill_restrictive(world, world.state, early_locations, early_rest_items, lock=True)
+        early_locations += early_priority_locations
+        fill_restrictive(world, world.state, early_locations, early_prog_items, lock=True)
```

Comment:
I wondered what the reason is for filling `early_rest_items` before `early_prog_items`, rather than the reverse order.

If we run out of space in sphere 1, we'll need progression items in sphere 1 to continue generating, right?

I'm imagining situations where `early_rest_items` take up all of sphere 1, and then no progression can happen.

Reply:
In an earlier version of this feature, it filled progression items into early priority locations first, then added non-progression items in and filled into non-priority locations.

However, this caused a guarantee that any first-sphere priority locations would contain items specified as early items.

I changed it to work like this so that I could avoid guaranteeing that early priority locations are filled first, without having to go back through unfilled locations to remove priority locations between fills